core/src/main/java/org/apache/struts2/components/template/FreemarkerTemplateEngine.java

```diff
@@ -144,15 +144,23 @@ public void close() throws IOException {
             }
         };
 
+        LOG.debug("Puts action on the top of ValueStack, just before the tag");
+        if (action == null) {
+            LOG.debug("Action from ActionInvocation is null, assuming action is on the top of ValueStack");
+            action = stack.pop();
+        }
         try {
             stack.push(templateContext.getTag());
+            stack.push(action);
             template.process(model, writer);
         } finally {
-            stack.pop();
+            stack.pop(); // removes action
+            stack.pop(); // removes tag
+            stack.push(action); // puts back action
```

Comment:
I meant this part. This lines only makes sense when `action` had been null previously. Otherwise I think it'll end up with a ValueStack which has duplicate actions in stack from a same instance.

Reply:
Good catch, fixed!